src/libsyntax_ext/deriving/mod.rs

```diff
@@ -74,3 +74,94 @@ fn call_intrinsic(cx: &ExtCtxt<'_>,
         span,
     }))
 }
+
+
+// Injects `impl<...> Structural for ItemType<...> { }`. In particular,
+// does *not* add `where T: Structural` for parameters `T` in `...`.
+// (That's the main reason we cannot use TraitDef here.)
+fn inject_impl_of_structural_trait(cx: &mut ExtCtxt<'_>,
+                                   span: Span,
+                                   item: &Annotatable,
+                                   deriving_sym: Symbol,
+                                   structural_path: generic::ty::Path<'_>,
+                                   push: &mut dyn FnMut(Annotatable)) {
+
+    // set the expn ID so we can use the unstable trait.
+    let span = span.fresh_expansion(syntax_pos::ExpnData::allow_unstable(
+        syntax_pos::ExpnKind::Macro(syntax_pos::MacroKind::Derive, deriving_sym),
+        span,
+        cx.parse_sess.edition,
+        [sym::structural_match][..].into()));
```

Comment:
`fresh_expansion` shouldn't be used here, instead `structural_match` should go into `#[allow_internal_unstable]` on the `macro PartialEq` and `macro Eq` definitions in libcore.

Reply:
done, thanks!